### ISSUE TITLE
64x Tfs 1.1 - Crash after 15 minutes

### ISSUE THREAD

**Issue by Fle-OTL** [orts/server#151]
Hey my server is crashing every 15 minutes when I have the npcs loaded, does anybody have this issue?

using orts datapack

**Comment by Schiffers** [comment]
you compiled in your computer?

**Comment by Fle-OTL** [comment]
yes I compiled in my computer

**Comment by Royalot** [comment]
64bit Version is bugged. Try the 32bit version

**Comment by EPuncker** [comment]
this is a very known issue here, so this is a duplicate or whatever you guys may call it as @TheSumm said in the other issue: it is low priority until we can really track down it

**Comment by Fle-OTL** [comment]
has there been any luck tracking the cause of crashes caused by npcsystem?

**Comment by WibbenZ** [comment]
Not really, but Mark said yesterday that it might be the npc lib files(https://otland.net/threads/the-forgotten-server-1-1.232064/#post-2236784)
So try to use the npc libs from the forgottenserver datapack, you will get alot of errors but it should still start. Then check how much ram it's using.

If it's the same then try to remove 10 npc files at a time till you have narrowed it down to a few npcs.

**Comment by rowaraujo** [comment]
Reload the NPCs is a burdensome task for the server.
Try updating everything and then restarted if possible.

**Comment by Schiffers** [comment]
how are the soluction to this problem? are crashing yet?

**Comment by Fle-OTL** [comment]
still crashing.. no luck finding the cause.

ninja said it may be the custom modules in the npc system..

I tried removing the dofile('data/npc/lib/npcsystem/customModules.lua')
from C:\Ot Project\tfs1.1\data\npc\lib -npc.lua 

But i get many errors when i do this and some npcs dont load, 

Also the server still crashes

I think this has too do with continuation passing style that is taking place without reason repeatedly eating memory until tfs crashes.

This is my unprofessional opinion!

-Martin

**Comment by Fle-OTL** [comment]
I think the reason the 32x is more stable, is because your memory is capped

However in 64x there is no memory cap and something is eating memory (some type of broken or outdated callback possibly?) I have no idea.

wish this were fixed.

**Comment by Schiffers** [comment]
"something eating memory" have a chance that have a inifite loop that are geting all memory and the tfs crash for memory end ?
@Fle-OTL you say about the 32x and 64x, but bro, how you will load the map in 32x ?
the server is for 64x to can load the map and all modules without problem

**Comment by ninjalulz** [comment]
This issue has been around since the start (late 2013(?)) of this project so I doubt there's anything wrong with the changes we have added recently.

I made a minor test with 200 NPCs (linked to the same script, default.lua) and default TFS libs (and ORTS libs). The memory usage by the server increased with either 4 or 8kb per second (and around 50kb with all NPCs, with ORTS libs at least).

Questions:
* How much memory is the server using when it crashes?
* The memory usage increases each second, but by how much?

**Comment by WibbenZ** [comment]
Done some more research:
Removed 489 NPCs (50% ~)
It goes now up 16kb insted of 30-50kb.

And if I use the other 341 (50% ~) NPCs that I removed before it goes up by 36 bytes.
I have now started going in alfabetical order:

A: They seem to leak at random intervals at 4 bytes per interval.

B, C, D, E, F, G, H, I, J, K: Seems stable, look like TFS is eating memory at startup and then releases it.
After a while it goes back down and is stable.

L: Is leaking, it takes a while but then starts leaking at 4kb / 5-10 sec.

Used the ORTS libs, TFS 64bit release (latest version) and the task manager to see if it leaks or not.
You should wait a while before you check if it leaked or not, as I said above TFS seems to be leaking abit in the beginning(after start) but then gets cleared after a while).
So let it idle for a couple of mins then check if it increases (wait 30-60 seconds).

**Comment by HyphenMaster** [comment]
This crash occurs only in Windows x64 ?
